specs/interop/token-bridging.md

```diff
@@ -2,96 +2,298 @@
-    _burn(msg.sender, amount);
+function sendERC20(address _to, uint256 _amount, uint256 _chainId, bytes memory _data) public returns (bool success) {
+  _burn(msg.sender, _amount);
+  bytes memory _message = abi.encodeWithSignature("finalizeSendERC20(address,uint256,bytes)", _to, _amount, _data);
+  _sendMessage(_chainId, _message);
+  return true
+}
 
-    L2ToL2CrossDomainMessenger.sendMessage({
-      _destination: chainId,
-      _target: address(this),
-      _message: abi.encodeCall(this.finalizeBridgeERC20, (to, amount))
-    });
-  }
+function finalizeSendERC20(address _to, uint256 _amount) external {
+  // this will be a modifier, inline here for clarity
+  require(msg.sender == address(L2ToL2CrossChainMessenger));
+  require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+  //
+  _mint(_to, _amount);
+}
 
-  function finalizeBridgeERC20(address _to, uint256 _amount) external {
-    require(msg.sender == address(L2ToL2CrossChainMessenger));
-    require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+function finalizeSendERC20(address _to, uint256 _amount, bytes memory _data) external {
+  require(msg.sender == address(L2ToL2CrossChainMessenger));
+  require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+  _mint(_to, _amount);
+  (bool success, ) = _to.call(_data);
+  require(success, "External call failed");
+}
 
-    _mint(to, amount);
-  }
+/////// Internal functions ///////////
+function _sendMessage(uint256 _destination, bytes memory _data) internal virtual {
+  L2ToL2CrossDomainMessenger.sendMessage(_destination, address(this), _data);
+}
+```
+
+Note: Some other naming options that were considered were `bridgeERC20`, `send()`, `xTransfer()`, `Transfer` (with a chainId parameter).
+
+## Invariants
+
+Besides the ERC20 invariants, the SuperchainERC20 will require the following interop specific properties:
+
+- Conservation of finalized `totalSupply`: The minted `amount` in `finalizeSendERC20()` should match the `amount` that was burnt in `sendERC20()`, as long as target chain has the initiating chain in the dependency set.
+  - Corollary 1: Cross-chain transactions will conserve the sum of `totalSupply` for each chain in the Superchain across safe blocks.
+  - Corollary 2: Each initiated but not finalized message (included in initiating chain but not yet in target chain) will decrease the `totalSupply` exactly by the burnt `amount`.
+  - Corollary 2: `SuperchainERC20s` should not charge a token fee or increase the balance when moving cross-chain.
+  - Question: what should we assume for chains not in the dependency set? a rollback method could modify this invariant to also consider that case.
+- Freedom of movement: Users should be able to send tokens into any target chain that has initiating chain in its dependency set.
+  - Question: should we add "deployed" to the invariant or assume its not going to be an issue?
```

Comment:
I think we could skip over the deployed idea here and have offchain tooling help to prevent footguns